Support install hosts via iPXE for virt and hana test

Julie-CAO opened the pull request:
- Schedule installation/ipxe_install.pm for tests with IPXE=1.
- Add more kernel and linuxrc arguments in bootscript for virt and hana tests.
- Finally don't touch the prefixing in autoyast profile path with SUT_IP in lib/autoyast.pm as it will be fixed in another PR. It breaks IPXE autoyast installation, fg. http://openqa.suse.de/tests/11712594#step/installation/1. 
- Prevent warnings `Use of uninitialized value...`, fg. https://openqa.suse.de/tests/11712594/logfile?filename=autoinst-log.txt
- "Boot flag" is not required anymore after vh015 firmwire upgrade
- Support passing "EXTRA_PXE_CMDLINE" to bootscripts.
- Enable sshd during autoyast installation for sake of debugging and post_fail_hook().
- Allow to install other SLE release with '+MIRROR_HTTP' as always.
- Be able to install TW with installer DVD outside O3. fg.[TW in my own openqa server](http://10.67.129.27/tests/77)
- Differentiate iPXE server on baremental-support service and static
  iPXE server by introducing a new setting, IPXE_STATIC(O3 and OSD share same ipxe menu)
- Add hard disk's wwns of two machines to autoyast profiles for virt test s in a seperate commit


Related ticket: https://progress.opensuse.org/issues/128060
Needles MR: https://gitlab.suse.de/openqa/os-autoinst-needles-sles/-/merge_requests/1647

Verification run: 
[xen tests with autoyast](http://10.67.129.27/tests/66)
[sle12sp5 with autoyast](http://10.67.129.27/tests/65)
[sle12sp5 xen without autoyast](http://10.67.129.27/tests/64)
[kvm without autoyast on OSD](https://openqa.suse.de/tests/11626483#)
[kvm with autoyast on OSD](https://openqa.suse.de/tests/11895292)
[prj2 tests with autoyast](http://10.67.129.27/tests/67)
[TW kvm test on O3](https://openqa.opensuse.org/tests/3522659)
[UEFI test with autoyast](http://10.67.129.27/tests/70)
[TW in my own openqa server](http://10.67.129.27/tests/77)

Dawei-Pang commented on tests/installation/ipxe_install.pm:
```diff
@@ -95,12 +99,26 @@ sub set_bootscript {
     if ($autoyast ne '') {
-        $cmdline_extra .= " autoyast=$autoyast ";
+	    #julie        $cmdline_extra .= " autoyast=$autoyast ";
+        $cmdline_extra .= " autoyast=$autoyast sshd=1 sshpassword=$testapi::password ";
```
During autoyast installation, no interactive is required, I think we need not sshd and sshpassword for boot cmdline, how about you?

varunkojha replied:
seems to depend on the trade-off between the potential security concerns of including passwords in the command line and the practical benefits of having SSH access for troubleshooting purposes during non-interactive AutoYaST installations. What happens if we remove it => A hook might need SSH access to collect important logs, I hope the understanding is correct.

Julie-CAO replied:
Yes, you get it right. This secrect of ssh login is not so sensible as the IPMI access.

waynechen55 replied:
Using ssh allows user to login anyway. I think we should keep it.

Julie-CAO replied:
https://en.opensuse.org/SDB:Linuxrc
```
ssh=1  --  This parameter enables access to linuxrc via SSH when performing the installation with YaST in text mode or via X11 forwarding. Use ssh -X root@hostname or ssh -Y root@hostname for X11 forwarding.

sshd=1  --- This parameter enables login to the installation system while the installation is running.
```

waynechen55 replied:
Yes. I was wondering whether ```ssh=1``` works with auotyast.

Julie-CAO replied:
Not sure. Let's see in future tests.  'ssh=1' is not propriate, because some tests also run on sol console or on VNC graphic mode.

Dawei-Pang commented on tests/installation/ipxe_install.pm:
```diff
@@ -15,7 +15,7 @@ use testapi;
 use ipmi_backend_utils;
 use version_utils qw(is_upgrade is_tumbleweed);
-use bootloader_setup 'prepare_disks';
+use bootloader_setup qw(prepare_disks bootmenu_default_params specific_bootmenu_params);
```
Shall we think about remove `bootmenu_default_params` `specific_bootmenu_params` in `use`, and do not touch `bootloader_setup.pm`?

Julie-CAO replied:
I am managing to append these arguments directly.

Dawei-Pang commented on lib/autoyast.pm:
```diff
@@ -713,9 +713,13 @@ sub upload_profile {
+    # fg. http://openqa.suse.de/tests/11712594#step/installation/1
+    # so comment out this part
+    # if (check_var('IPXE', '1')) {
+    #     $path = get_required_var('SUT_IP') . $path;
```
Reproduced the issue when including IPXE=1, it generates the `$path` likes`${SUT_IP}autoyast_hanaperf/mu.xml`, but no code loading autoinst.xml following it.
I agree to remove this code block.

alice-suse replied:
Will it change the path for kernel team? Better to involve kernel team PO for review when PR is ready.

Julie-CAO replied:
Finally I made this change limited to virt tests and hana tests only.

Julie-CAO commented on tests/installation/ipxe_install.pm:
```diff
@@ -95,12 +99,26 @@ sub set_bootscript {
     $cmdline_extra .= " root=/dev/ram0 initrd=initrd textmode=1" if check_var('IPXE_UEFI', '1');
```
Ok. done.

Dawei-Pang replied:
Thanks!

Julie-CAO commented on tests/installation/ipxe_install.pm:
```diff
@@ -97,10 +101,23 @@ sub set_bootscript {
 
+    # Extra options for virtualization tests with ipmi backend
+    if (check_var('VIDEOMODE', 'text')) {
+        my $serial_dev = get_var('SERIALDEV', 'ttyS1');
```
There is no CONSOLE but SEREAL_DEV in virt tests. I am thinking of change to use CONSOLE or support both of them.

Dawei-Pang replied:
I suggest supporting both of them, thanks!

Julie-CAO replied:
yes. find that `SEREAL_DEV` is a mandatory setting for virt tests, so support both of them.

Julie-CAO committed:
Add disk device ID for 3 Beijing machines

mdoucha commented on tests/installation/ipxe_install.pm:
```diff
@@ -48,7 +48,7 @@ sub poweron_host {
     while (1) {
         my $stdout = ipmitool('chassis bootparam get 5');
-        last if $stdout =~ m/Boot Flag Valid[\d\D]*Force PXE/;
+        last if $stdout =~ m/Force PXE/;
```
If you drop `Boot Flag Valid` from the regex, the function will accept a state where IPMI boot override is disabled and the SUT will boot according to BIOS settings, ignoring the `Force PXE` flag.

Julie-CAO replied:
To check `Boot Flag Valid` was added by me in a previous PR because we had a machine that only worked with this flag. We upgraded the BIOS and firmwire of this machine days ago, after that I found Boot Flag always remains 'Invalid' which stopped the test to be continued, fg. http://10.67.129.27/tests/28/logfile?filename=autoinst-log.txt

And it seems that the flag does not effect the tests anymore. I will do more verification runs on more machines.

mdoucha replied:
OK, never mind then.

Julie-CAO replied:
wait a moment... I might speak too early. it seems that something wrong with the machine, it boot from pxe to install SLE repeatedly...   I am going to figure out why its Boot Flag keeps 'Invalid' firstly.

alice-suse commented on data/virt_autotest/host_unattended_installation_files/autoyast/dev_host_15.xml.ep:
```diff
@@ -118,7 +118,7 @@
   </ntp-client>
   <partitioning config:type="list">
-  % my $wwn = {'quinn' => 'wwn-0x5000c50099db2117', 'kermit-1' => 'wwn-0x500a075119406ab6', 'gonzo-1' => 'wwn-0x500a075119406aa6', 'fozzie-1' => 'wwn-0x55cd2e414f1f16f1', 'scooter-1' => 'wwn-0x55cd2e414f1760e2', 'amd-zen3-gpu-sut1-1' => 'wwn-0x500a075133755d4b', 'ix64ph1075' => 'wwn-0x5000c5004f25d745', 'openqaipmi5' => 'wwn-0x5000c5008711f2fc'};
+	  % my $wwn = {'quinn' => 'wwn-0x5000c50099db2117', 'kermit-1' => 'wwn-0x500a075119406ab6', 'gonzo-1' => 'wwn-0x500a075119406aa6', 'fozzie-1' => 'wwn-0x55cd2e414f1f16f1', 'scooter-1' => 'wwn-0x55cd2e414f1760e2', 'amd-zen3-gpu-sut1-1' => 'wwn-0x500a075133755d4b', 'ix64ph1075' => 'wwn-0x5000c5004f25d745', 'openqaipmi5' => 'wwn-0x5000c5008711f2fc', '10.67.130.52' => 'wwn-0x5000c500e28a91bb', '10.67 ...
```
Why not use domain name here? Better than direct ip?

Julie-CAO replied:
yes domaine name is better. 

With autoyast installation, we must change the settings in workers.ini if `SUT_IP=10.67.130.15` to `SUT_IP=vh015.qa2.suse.asia`.

Julie-CAO replied:
I added this task to https://progress.opensuse.org/issues/134261

alice-suse commented on tests/installation/ipxe_install.pm:
```diff
@@ -87,19 +87,44 @@ sub set_bootscript {
+    if (get_var('VIRT_AUTOTEST') || get_var('HANA_PERF')) {
+        $cmdline_extra .= " vga=791 video=1024x768 vt.color=0x07";
+        my $serial_dev;
+        if (get_var('IPXE_CONSOLE')) {
```
What's the background for differentiating IPXE_CONSOLE and SERIALDEV? It seems both are to speciy the real serial console?

Julie-CAO replied:
yes, IPXE_CONSOLE was introduced by ipxe installation and is being used by kernel team. SERIALDEV is being used by virt team and it is mandatory as it is referenced in host autoyast profiles. So we have to support both of them.

Julie-CAO replied:
usually these value are valid for them:
```
IPXE_CONSOLE=ttyS1,115200
SERIALDEV=ttyS1
```

alice-suse replied:
ok, maybe adding a comment to explain them will be useful for reference

alice-suse commented on tests/installation/ipxe_install.pm:
```diff
@@ -141,6 +171,21 @@ END_BOOTSCRIPT
 
+sub enter_o3_ipxe_boot_entry {
+    ipmitool('chassis power reset') unless check_screen([qw(o3-ipxe-menu ipxe-boot-failure)], 180);
+    assert_screen('o3-ipxe-menu', 300);
```
With default TIME_SCALE 3, it wil give 900 seconds(15 minutes) to asset ipxe menu. Is it really needed? How about fail earlier?

Julie-CAO replied:
it depends on the restart time including power off and power on. I cannot get an acurate time from a test case, https://openqa.opensuse.org/tests/3446335#step/ipxe_install/22, but the time seems to be shorter than 900s. Let's reduce to 210 and keep eyes on it in future tests.

alice-suse commented on tests/installation/ipxe_install.pm:
```diff
@@ -156,27 +201,24 @@ sub run {
         else {
-            record_info("TW install?", "Pls make sure it is Tumbleweed that is being installed.", result => 'softfail');
+            record_info("Install others?", "Pls make sure the product that is expected to be installed.");
         }
```
Do we need to add the  needles per sles product? If not, that's fine. Otherwise maybe a little expensive just for record_info.

Julie-CAO replied:
I have added needles for TW and SLE seperately, no need to add needles for each SLE release.

alice-suse commented:
Thanks @Julie-CAO  for the great work. I left only some minor comments.

frankenmichl commented on lib/autoyast.pm:
```diff
@@ -713,9 +713,13 @@ sub upload_profile {
     my $path = $args{path};
 
-    if (check_var('IPXE', '1')) {
-        $path = get_required_var('SUT_IP') . $path;
```
We should agree with @mdoucha if we want to keep this or not, or maybe add another variable to control this?

Julie-CAO replied:
I see it is moved to another place in https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/17583. I'll not leave this part to your team as you are working on it. In my a few verify tests, these lines are not required at here.

Julie-CAO replied:
I keep this change in this PR for a moment for sake of verifying tests on openqa.

Julie-CAO commented:
Hello everyone, the code change is ready. The descripton and verification runs are updated. Welcome review! @frankenmichl @mdoucha @Dawei-Pang @alice-suse @guoxuguang @waynechen55 @nanzhg @tonyyuan1 @RoyCai7 @tbaev @varunkojha @JoyceNa @sudarshannm @rmaliska

waynechen55 commented:
Does this support aarch64 as well ?

waynechen55 commented on tests/installation/ipxe_install.pm:
```diff
@@ -87,19 +95,46 @@ sub set_bootscript {
+
+    # Extra options for virtualization tests with ipmi backend
+    if (get_var('VIRT_AUTOTEST') || get_var('HANA_PERF')) {
+        $cmdline_extra .= " vga=791 video=1024x768 vt.color=0x07";
```
Some options, like ```vga``` is already deprecated.

Julie-CAO replied:
Are these 3 parameters all deprecated? I am not aware of it. I'll do some reseach and tests before removing them.

Julie-CAO replied:
Are these 3 parameters all deprecated? I am not aware of it. I'll do some reseach and tests before removing them.

Julie-CAO replied:
`vga=xx` is deprecated in 32-bit linux and Grub2. `video=xx` and `vt.colar=xx` are still in use.  I removed `vga=791`.

Julie-CAO commented:
> Does this support aarch64 as well ?

Aarch64 boot with UEFI instead of legacy BIOS? UEFI should be supported, but about aarch64 I am not sure. I guess there should be a new ticket to handle it seperately.

waynechen55 commented on lib/main_common.pm:
```diff
@@ -425,6 +425,9 @@ sub load_boot_tests {
     }
+    elsif (get_var("IPXE")) {
+        loadtest "installation/ipxe_install";
+    }
```
So ```ipxe_install``` has never been scheduled before in ```main_common.pm``` ?

Julie-CAO replied:
No, it is usually loaded in job scheduler yaml files directly in test suites.

waynechen55 commented on tests/installation/ipxe_install.pm:
```diff
@@ -70,11 +70,19 @@ sub set_bootscript {
-    $console =~ s/^\s+|\s+$//g;
+    $autoyast =~ s/^\s+|\s+$//g if $autoyast;
+    $regurl =~ s/^\s+|\s+$//g if $regurl;
+    $console =~ s/^\s+|\s+$//g if $console;
```
I do not think these three ```if``` are very necessary. Keep these line as what they were is enough.

Julie-CAO replied:
They bring `uninitialized value ...` warnings in auto-inst.txt. To remove this kind of warning and make logs clean follows https://progress.opensuse.org/issues/120750.

waynechen55 replied:
```$autoyast``` already initialized by ```get_var('AUTOYAST', '');```. So I think you just need
```
my $regurl = get_var('SCC_URL', '');
my $console = get_var('IPXE_CONSOLE', '');
```
instead of using ```if```.

Julie-CAO replied:
Yes, it works. thanks.

waynechen55 commented:
I think you need some verifications with developing product, namely 15-SP6 to verify it also works with developing product.

waynechen55 commented on tests/installation/ipxe_install.pm:
```diff
@@ -141,50 +181,54 @@ END_BOOTSCRIPT
 
     #virtualization tests use a static ipxe configuration file in O3
-    set_bootscript unless get_var('VIRT_AUTOTEST') && is_tumbleweed;
+    set_bootscript unless get_var('VIRT_AUTOTEST') and is_tumbleweed and get_var('OPENQA_URL') =~ "openqa.opensuse.org";
```
According to your comment, is_tumbleweed should be removed from ```unless``` ?

Julie-CAO replied:
I don't know if ALP behaves same way as Tumbleweed, so just to be limited to TW.

mdoucha replied:
This should be `set_bootscript unless get_var('IPXE_STATIC'); or something like that. Then set `IPXE_STATIC` whereever it's needed. Checking `DISTRI` or `OPENQA_URL` is really terrible approach.

frankenmichl replied:
And please always keep in mind the baremetal support service has no persistency for it's data. If, for any reason, the service restarts, it will lose all the static configuration. Would be a better solution to always set your bootscript if you don't want to risk boot failures in that case.

Julie-CAO replied:
> And please always keep in mind the baremetal support service has no persistency for it's data. If, for any reason, the service restarts, it will lose all the static configuration. Would be a better solution to always set your bootscript if you don't want to risk boot failures in that case.

Thanks. There is no baremetal support service deployed or be accessible on O3. There is only a static iPXE menu shared by tests.

frankenmichl replied:
Okay! If you need the service on o3 just reach out, I bet we can do something with that.

Julie-CAO replied:
Thank you, Michael. I did some research before my tests are deployed in O3 with iPXE. The dhcp server in O3 is dnsmasq, it seems that the http redirection for designated machines are not supported. All the tests have to share the same iPXE configurations. In addition, accessing internal SUSE network(IP likes 10.xx) is not allowed on O3. So finally we choosed the static iPXE menu.

Julie-CAO replied:
> This should be `set_bootscript unless get_var('IPXE_STATIC');` or something like that. Then set `IPXE_STATIC` whereever it's needed. Checking `DISTRI` or `OPENQA_URL` is really terrible approach.

I noticed that OSD began to use the same static iPXE menu on O3. `get_var('IPXE_STATIC')` is more approriate. thanks.

Julie-CAO commented:
> 1. I think you need some verifications with developing product, namely 15-SP6 to verify it also works with developing product.
Verifiying 15-SP6 will be done during it is deployment. They are seperate tasks. Introduing unmature sp6 tests will bring unexpected troubles. It is not smart.

> 2. What a failed ipxe install looks like ? Enough information available for user to do follow-up ?
Here is a recent failed case with incorrect kernel path: http://10.67.129.27/tests/85
screenshots + logs + videos will help locating failures.

Julie-CAO committed:
Support ipxe installation for ipmi backend tests(virt, perf and hana tests)

- Add more kernel and linuxrc arguments in bootscript for virt and hana tests.
- Schedule installation/ipxe_install.pm for tests with IPXE=1.
- Support passing EXTRA_PXE_CMDLINE to bootscripts.
- Allow to install other SLE release with '+MIRROR_HTTP' as always.
- Be able to install TW with installer DVD outside O3.
- Differentiate iPXE server on baremental-support service and static
  iPXE server by introducing a new setting, IPXE_STATIC
- 'Boot flag' is not required anymore after vh015 firmwire upgrade
- Prevent warnings Use of 'Uninitialized value...'

mdoucha commented on tests/installation/ipxe_install.pm:
```diff
@@ -75,6 +75,14 @@ sub set_bootscript {
+    # Install other release rather than the VERSION
+    if (get_var('REPO_0_TO_INSTALL')) {
+        my $openqa_url = get_required_var('OPENQA_URL');
+        $openqa_url = 'http://' . $openqa_url unless $openqa_url =~ /http:\/\//;
```
This condition will result in `$openqa_url = 'http://https://openqa.opensuse.org'`

Julie-CAO replied:
Yes, good catch. thanks. fixed.

mdoucha replied:
The correct fix is this:

    $mirror_http = autoinst_url("/assets/repo/" . get_var("REPO_0_TO_INSTALL"));

mdoucha replied:
But more importantly, why do you even add the `REPO_0_TO_INSTALL` variable to do the exact same thing as `MIRROR_HTTP`?

Julie-CAO replied:
> The correct fix is this:
> 
> ```
> $mirror_http = autoinst_url("/assets/repo/" . get_var("REPO_0_TO_INSTALL"));
> ```

Just tried it out, but the path is not correct. It is the job space in worker instead of the http path in openqa repo.

```
https://openqa.suse.de:20523/T8bHv8ZC0urZFd92/assets/repo/fixed/SLE-12-SP5-Server-DVD-x86_64-GM-DVD1

```

I need:
```
https://openqa.suse.de/assets/repo/fixed/SLE-12-SP5-Server-DVD-x86_64-GM-DVD1
```

mdoucha replied:
Ah, right. Normal `REPO_N` would be copied to the worker and then accessed via `autoinst_url()`. But the `_TO_INSTALL` suffix prevents that here.

On the other hand, when the worker is physically located far away from the OpenQA server, this approach will stress the network a lot. But we also have several mirror hosts located near workers which are supposed to be used throught the `MIRROR_HTTP` variable. Why not use that?

Julie-CAO replied:
`MIRROR_HTTP` is for installing the current products, fg.sle-15-SP5. but in virt tests, we need install different products, such as sle-12-sp5, sle15-sp4. For these group of tests, `REPO_0_TO_INSTALL` is introduced to flag which product will be installed.

The baremetal machines locate in different locations and VLAN.  medium in openqa server is a good place to access, and it is easier to maintain.

mdoucha replied:
> `MIRROR_HTTP` is for installing the current products, fg.sle-15-SP5. but in virt tests, we need install different products, such as sle-12-sp5, sle15-sp4. For these group of tests, `REPO_0_TO_INSTALL` is introduced to flag which product will be installed.

All repos available on OSD are also available on the dist mirrors, e.g. sle12-sp5 installer repo you've used as an example above (SLE-12-SP5-Server-DVD-x86_64-GM-DVD1) would be `http://[nearest dist mirror]/install/SLP/SLE-12-SP5-Server-GM/x86_64/DVD1/`. In the worst case, you can set `MIRROR_HTTP` to the OSD repo URL.

> The baremetal machines locate in different locations and VLAN. medium in openqa server is a good place to access, and it is easier to maintain. `REPO_0_TO_INSTALL=fixed/SLE-xx` has been widely used in virt tests, code change should be compitible with all existing tests.

You can set `MIRROR_HTTP` in the medium like this:

    MIRROR_HTTP=http://%MIRROR_SERVER%/install/SLP/SLE-12-SP5-Server-GM/x86_64/DVD1/

and then set `MIRROR_SERVER` to the preferred mirror hostname on each worker.

Julie-CAO replied:
We cannot set MIRROR_HTTP in the medium for each test suite because they are in same job group and run with same medium. All the tests are triggered at same time with same MIRROR_HTTP in one medium. ie. We install sle12sp5 in a sle15sp6 job group.  As far as I know, MIRROR_HTTP is passed in the command to trigger an entire job group(or even a product), it will override the configurations in test suite level.

mdoucha replied:
You can set `+MIRROR_HTTP=...` in the testsuite or job entry in the jobgroup YAML. The `+` will override the setting from medium and scheduled product.

mdoucha replied:
See for example the ALP config in https://openqa.suse.de/admin/job_templates/498 where `+HDD_1` is used to override `HDD_1` set by `isos post`.

Julie-CAO replied:
Yes, `+MIRROR_HTTP=` works well in virt tests with non-developing versions. In addition, setting it in test suites is more flexable. fg. http://10.67.129.27/tests/94

Dawei-Pang commented on tests/installation/ipxe_install.pm:
```diff
@@ -124,6 +159,11 @@ END_BOOTSCRIPT
 
     my $response = HTTP::Tiny->new->request('POST', $url, {content => $bootscript, headers => {'content-type' => 'text/plain'}});
     diag "$response->{status} $response->{reason}\n";
```
Suggest to add the code as following to checking if IPXE_HTTPSERVER works
```    die "Fail to post IPXE_HTTPSERVER\n" unless $response->{success};```

Julie-CAO replied:
Added. Trying out to verify...

mdoucha replied:
This is fixed in #17583. We could merge that first and rebase this on top.

Julie-CAO replied:
Sounds great. I remove the change from my PR.

waynechen55 commented:
> > 1. I think you need some verifications with developing product, namely 15-SP6 to verify it also works with developing product.
> 
> Verifiying 15-SP6 will be done during its deployment. They are seperate tasks. Introducing unmature sp6 tests will bring unexpected troubles. It is not smart.

I do not think this is ready to be merged without verification with 15-SP6.

Julie-CAO commented:
the new verification run: https://openqa.opensuse.org/tests/3537864#details

PR is ready. Please help merge.

alice-suse commented:
@mdoucha  Hi Martin, there is change request from you. Is it old one or active one? I am considering to merge if it is resolved.

alice-suse reviewed (approved):
LGTM

mdoucha commented on tests/installation/ipxe_install.pm:
```diff
@@ -141,50 +173,54 @@ END_BOOTSCRIPT
+    # set limit to 10 in case some keys don't go through
+    for (1 .. 10) {
+        last if check_screen([qw(load-linux-kernel load-initrd)], 3);
+        send_key "ret";
```
Is this logic correct? This will keep sending the "Enter" keypress blindly even if no dialog shows up.

varunkojha replied:
@mdoucha do we need a timeout?

my $max_attempts = 10;
my $timeout = 3;

for (1 .. $max_attempts) {
    if (check_screen([qw(load-linux-kernel load-initrd)], $timeout)) {
        last;
    }
    send_key "ret";
}

varunkojha replied:
3 secs is already there

mdoucha replied:
No. I think the logic should look like this:

    send_key "ret" if check_screen("needle-checking-whether-the-dialog-showed-up", 30);

The jobs using the ipxe_install module usually run with 3x timeout multiplier so the key will be sent over up to 90 seconds. If the dialog doesn't show up and interactive YaST starts quickly, this'll end up advancing the installation several steps and breaking subsequent installation modules.

Julie-CAO replied:
this logic was introduced by https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/17571. It is to adapt the new iPXE menu. the new iPXE menu need user's input to to start the command. "ret" will be sent until it is confirmed that the bootscript has been executed.

see this screenshot: https://openqa.opensuse.org/tests/3537864#step/ipxe_install/6

"ret" keys are required in the 'edit' mode, actually finally 3 "ret" keys are typed in this test.

```
[2023-08-31T11:28:52] no match: -0.7s, best candidate: ipxe_install-o3-ipxe-menu-20220811 (0.00)
[2023-08-31T11:28:52] >>> testapi::_check_backend_response: match=o3-ipxe-menu timed out after 0 (check_screen)
[2023-08-31T11:28:52] tests/installation/ipxe_install.pm:209 called ipxe_install::enter_o3_ipxe_boot_entry -> tests/installation/ipxe_install.pm:186 called testapi::check_screen
[2023-08-31T11:28:52] <<< testapi::check_screen(mustmatch=[
    "load-linux-kernel",
    "load-initrd"
  ], timeout=3)
[2023-08-31T11:28:53] no match: 8.2s, best candidate: ipxe_install-load-initrd-20220629 (0.00)
...
[2023-08-31T11:29:02] no match: -0.8s, best candidate: ipxe_install-load-linux-kernel-20220629 (0.79)
[2023-08-31T11:29:021] >>> testapi::_check_backend_response: match=load-initrd,load-linux-kernel timed out after 9 (check_screen)
[2023-08-31T11:29:02] tests/installation/ipxe_install.pm:209 called ipxe_install::enter_o3_ipxe_boot_entry -> tests/installation/ipxe_install.pm:187 called testapi::send_key
[2023-08-31T11:29:02] <<< testapi::send_key(key="ret", wait_screen_change=0)
[2023-08-31T11:29:03] tests/installation/ipxe_install.pm:209 called ipxe_install::enter_o3_ipxe_boot_entry -> tests/installation/ipxe_install.pm:186 called testapi::check_screen
[2023-08-31T11:29:03] <<< testapi::check_screen(mustmatch=[
    "load-linux-kernel",
    "load-initrd"
  ], timeout=3)
[2023-08-31T11:29:03] no match: 8.7s, best candidate: ipxe_install-load-initrd-20220629 (0.00)
...
[2023-08-31T11:29:12] no match: -0.3s, best candidate: ipxe_install-load-linux-kernel-20220629 (0.79)
[2023-08-31T11:29:12] >>> testapi::_check_backend_response: match=load-initrd,load-linux-kernel timed out after 9 (check_screen)
[2023-08-31T11:29:12] tests/installation/ipxe_install.pm:209 called ipxe_install::enter_o3_ipxe_boot_entry -> tests/installation/ipxe_install.pm:187 called testapi::send_key
[2023-08-31T11:29:12] <<< testapi::send_key(key="ret", wait_screen_change=0)
[2023-08-31T11:29:13] tests/installation/ipxe_install.pm:209 called ipxe_install::enter_o3_ipxe_boot_entry -> tests/installation/ipxe_install.pm:186 called testapi::check_screen
[2023-08-31T11:29:131] <<< testapi::check_screen(mustmatch=[
    "load-linux-kernel",
    "load-initrd"
  ], timeout=3)
...
[2023-08-31T11:29:22] tests/installation/ipxe_install.pm:209 called ipxe_install::enter_o3_ipxe_boot_entry -> tests/installation/ipxe_install.pm:187 called testapi::send_key
[2023-08-31T11:29:22] <<< testapi::send_key(key="ret", wait_screen_change=0)
[2023-08-31T11:29:23] tests/installation/ipxe_install.pm:209 called ipxe_install::enter_o3_ipxe_boot_entry -> tests/installation/ipxe_install.pm:186 called testapi::check_screen
[2023-08-31T11:29:23] <<< testapi::check_screen(mustmatch=[
    "load-linux-kernel",
    "load-initrd"
  ], timeout=3)
[2023-08-31T11:29:23] no match: 8.7s, best candidate: ipxe_install-load-initrd-20220629 (0.00)
[2023-08-31T11:29:25] no match: 6.7s, best candidate: ipxe_install-load-initrd-20220629 (0.00)
[2023-08-31T11:29:26] >>> testapi::_handle_found_needle: found ipxe_install-load-initrd-20220629, similarity 1.00 @ 87/386
```

Julie-CAO replied:
the logic works now. Let's fix it in another PR when we find it bring troubles.

alice-suse replied:
It seems to me @mdoucha comment is more about accurately handle that operating of sending ret and make test more efficient than blindly trying, which makes sense IMO. @Julie-CAO Do you think this can be improved in future PR by using some needle like needle-checking-whether-the-dialog-showed-up? To me, this is a good thing but minor and not directly related with the main scope of the PR, better to merge it first and improve that later.

If no against comment today, will merge first.

Julie-CAO replied:
Yes, this part was about the new iPXE menu just introduced days ago. It is out of the scope of my PR. And it involves the tests in O3 only. I don't have plan to improve it in this PR.

About if it will be improved in the future, we need monitor more tests before give a solution, because 3 more needles("needle-checking-whether-the-dialog-showed-up")  from SOL console are required. If it is more efficient actually, we need more considerations.

mdoucha replied:
OK, future PR is good enough.

waynechen55 commented:
> 

@Julie-CAO Could you help make a verification run with 15-SP6, build of which just works as normal.

Julie-CAO commented:
> > 
> 
> @Julie-CAO Could you help make a verification run with 15-SP6, build of which just works as normal

sle15sp6 with autoyast would fail until https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/17583 is merged.

waynechen55 commented:
> > > 
> > 
> > 
> > @Julie-CAO Could you help make a verification run with 15-SP6, build of which just works as normal
> 
> sle15sp6 with autoyast would fail until #17583 is merged.

If this is the case, I do not think we need to rush merge this pull request. This should only be merged after pr#17583, otherwise, it will block all test runs.

Julie-CAO commented:
> If this is the case, I do not think we need to rush merge this pull request. This pull request should only be merged after [pr#17583](https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/17583), otherwise it will block all test runs. After [pr#17583](https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/17583) is merged, you can help do a verification with 15-SP6 to make sure merging this pull request will not block test runs.

No, it won't block sle15sp6 tests with this PR merged because the machines are still booting from PXE on OSD. 

That PR and my PR edited the same file, it would be better that my PR go in first and that PR go in next, and my needles PR got merged too. Finally we switch the workers on OSD to iPXE.

waynechen55 commented:
> > If this is the case, I do not think we need to rush merge this pull request. This pull request should only be merged after [pr#17583](https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/17583), otherwise it will block all test runs. After [pr#17583](https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/17583) is merged, you can help do a verification with 15-SP6 to make sure merging this pull request will not block test runs.
> 
> No, it won't block sle15sp6 tests with this PR merged because the machines are still booting from PXE on OSD.
> 
> That PR and my PR edited the same file, it would be better that my PR go in first and that PR go in next, and my needles PR got merged too. Finally we switch the workers on OSD to iPXE.

I think you can pick a worker later to have a try with iPXE and autoyast first before full depoyment.

Julie-CAO commented:
> > No, it won't block sle15sp6 tests with this PR merged because the machines are still booting from PXE on OSD.
> > That PR and my PR edited the same file, it would be better that my PR go in first and that PR go in next, and my needles PR got merged too. Finally we switch the workers on OSD to iPXE.
> 
> I think you can pick a worker later to have a try with iPXE and autoyast first before full depoyment.

Thanks. we had a pilot worker, scooter, is booting from iPXE. It is out of our worker pool now.

alice-suse commented:
Thanks  all for the review, merge it now.
